narciuch/assignment2/src/TreeAncestors.java

```diff
@@ -0,0 +1,135 @@
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+public class TreeAncestors<T> {
+
+  public void printAncestors(T key, Node treeRoot) {
+    List<Node> result = getAncestors(key, treeRoot);
+    printNodeList(result);
+  }
+
+  /**
+   * Returns a list of nodes that are parents, grandparents etc, of the node
+   * with the value of 'key'.
+   *
+   * @param key : The value of the node whos ancestors you want to find
+   * @param node : The root of the tree in which you want to search for the ancestors
+   * @return : List of nodes which are the ancestors of the node with value 'key'.
+   *           These are in the order of parent, grandparent, ..., root
+   */
+  public List<Node> getAncestors(T key, Node node) {
+    if (node == null) {
+      return Collections.emptyList();
+    }
+    List<Node> path = new ArrayList<>();
+    boolean result = getAncestorsHelper(key, node, path);
+    Collections.reverse(path); //Puts the ancestors into the desired order
+    return path;
+  }
+
+  /**
+   * Used to find the path from the root of a tree to a particular node
+   * (i.e. the nodes ancestors)
+   *
+   * @param key : The value of the desired node
+   * @param node : Tree to search through
+   * @param path : Indicates the visited nodes in the path from the original root directly to
+   *             the current 'node'
+   * @return true if the node with value 'key' was found in the tree 'node'
+   */
+  public boolean getAncestorsHelper(T key, Node node, List<Node> path) {
+    if (node == null) {
+      return false;
+    }
+    if (node.getValue() == key) {
+      return true;
+    }
+    boolean left = false;
+    boolean right = false;
+    if (node.getLeft() != null) {
+      path.add(node);
+      left = getAncestorsHelper(key, node.getLeft(), path);
+    }
+    if (node.getRight() != null) {
+      path.add(node);
+      right = getAncestorsHelper(key, node.getRight(), path);
+    }
+    if (path.size() != 0 && !left && !right) {
+      node = path.get(path.size() - 1); //Take most recently added node
+      path.remove(path.size() - 1);
```

Comment:
Optional: would it be possible to reorganize code so that elements are added to the path only when necessary rather than always added and then often removed?